util/base_parser.h

```diff
@@ -32,6 +32,15 @@ class BaseParser {
 
     constexpr void advance(std::size_t n) { pos_ += n; }
 
+    constexpr void back(std::size_t n) { pos_ -= n; }
+
+    constexpr void reset() { pos_ = 0; }
+
+    constexpr void reset(std::string_view input) {
```

Comment:
What's the use-case for these reset and back member functions? So far we've only used this for throwaway parsers, e.g.
`CssStyleSheet css = CssParser::parse("html { font-size: 1em; }");` where `parse(char const *)` internally creates a parser, creates a stylesheet, and returns it.

Reply:
There's plenty of places in the WHATWG URL standard where they describe the parser going backward. Ctrl+F "decrease pointer". Same with starting over from the beginning of the parser input.

Reply:
Starting over from the beginning is sad, but fair enough, that's how the spec goes. :P

I'm mostly wondering about `reset(std::string_view)`, but I guess maybe you create a `std::string` for the `UrlParser` to keep alive and pass that to reset?

Reply:
Yup, exactly.